[2단계 - 자동차 경주 리팩터링] 다온(조형익) 미션 제출합니다.

ikjo39 opened the pull request:
안녕하세요 리뷰어님 6기 다온입니다.
2단계 미션도 아래에 정리해보았습니다.

## 미션을 수행하며 진행된 일

Model과 View의 분리는 1단계 미션에서 고민하였기에 다시 점검하였습니다. 이로 인한 변경점은 없습니다!
그래서 1단계의 피드백을 중점으로 리팩토링하였습니다.

---

## 코드 작성시 신경 쓴 점

- '**VO에서 Validation 을 적용하여도 문제가 없을까?**' 란 고민을 먼저 한 후 코드에 적용해보았어요.
  - VO 내 데이터의 불변성을 해치지 않으므로 적용해도 된다는 결론을 내렸습니다.
- TryCount 역시 VO로 변경하였습니다.
  - 이 역시 VO로 관리하였을 때 더 이점이 있다 생각하였습니다.

---

## 질문

- Position 객체 내 **차량의 움직임**을 구현하기 위해 아래 메서드를 VO에 추가하였는데요.
  - VO의 성격을 해치지 않았다 생각하지만 불필요한 객체 생성으로 메모리를 너무 낭비하진 않는지 찝찝함이 남아있습니다.
```java
public class Position {
    //... 
    public Position move() {
        return new Position(position + 1);
    }
```

- 리팩토링을 하며 이 과정을 기능 요구사항 문서처럼 남겨둬야하나? 라는 의문이 생겼는데요.
  - 해야한다면 문서작성과 리팩토링의 순서 또한 궁금합니다 ㅎ  (선 문서 작성, 후 리팩토링 or 선 리팩토링, 후 문서 작성)
  - 추가로 현업에서는 기능 요구사항 정의서, 테스트 코드 이외에 또 어떤 문서들을 작성하는지 궁금해요! (꽤 있겠지만...)

Commit:
feat: 예외메세지 내 입력값 표시 추가

Commit:
feat: 가장 멀리 전진한 위치 산출 로직 리팩토링

Commit:
refactor: Car 내 필드 VO로 리팩토링

- 원시값 필드 -> VO로 변경
- 차 이름 Validation CarName 내로 이동
- CarName 테스트 코드 추가
- @BeforeEach 활용하여 테스트 코드 수정

Commit:
refactor: RacingGame 내 TryCount VO로 변경

- TryCount VO 추가
- Validation 로직 TryCount로 이동
- 관련 테스트 코드 수정

Commit:
refactor: 경주 결과 출력 로직 리팩토링

- 가독성을 위한 변수 분리
- 예상 출력 문구 주석 추가

Commit:
Style: 불필요한 변수 인라인화

sihyung92 commented on src/main/java/racingcar/controller/RacingController.java:
```diff
@@ -32,14 +34,14 @@ public void run() {
     private RacingGame createRacingGame(Cars cars) {
-        int tryCount = inputView.readTryCount();
-        return new RacingGame(tryCount, cars, new RandomNumberGenerator());
+        return new RacingGame(new TryCount(inputView.readTryCount()), cars, new RandomNumberGenerator());
```
전 수정하기 이전의 코드가 더 가독성있다고 생각합니다 ㅎ 
라인수를 줄이는 것 보다 늘리는 게 더 나을때도 있더라구요

ikjo39 replied:
제가 계속 읽다보니 익숙해져서 괜찮다 싶었나봐요.
수정한 코드는 오른쪽으로 읽어가며 해석해야하니 이해하는데 시간이 더 걸리겠네요 😅

sihyung92 commented on src/main/java/racingcar/model/Car.java:
```diff
@@ -2,48 +2,34 @@
 
```
model pacakge의 커버리지가 높네요 👍
테스트 실행 옵션 중 커버리지를 확인할 수 있는 옵션이 있습니다.
<img width="796" alt="image" src="https://github.com/woowacourse/java-racingcar/assets/51393021/c9d07c34-03b8-434a-95e6-f88a42b7f05e">

ikjo39 replied:
오오 이렇게 유용한 기능이 있었네요!! 자주 활용해봐야겠어요

sihyung92 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -22,14 +22,14 @@ public void makeCarMove(RandomNumberGenerator generator) {
     public RoundResult getRoundResult() {
```
makeCarMove() 메서드가 RoundResult를 반환하는 방향은 어떤가요?

getRoundResult()는 makeCarMove() 메서드와 시간적 의존을 가지고 있는 메서드입니다. (다음 makeCarMove()가 실행되기 전 실행되어야 함)
usecase에서도 항상 함께 쓰이고 있는데, 하나의 메서드로 합쳐 제공한다면 시간의존으로 인한 버그 (makeCarMove를 먼저 두번하고 roundResult를 두번 호출한다던지)를 막을 수 있을거 같아요.

sihyung92 replied:
그리고 RandomGenerator의 랜덤성으로 인해 makeCarMove()에 대한 단위테스트가 없는 문제 (RacingGame의 run 테스트로 인해 커버가 된 것처럼 보이지만, 사실은 단위테스트가 존재하지 않음)도 함께 해결할 수 있어 1석2조 같습니다

ikjo39 replied:
저는 RoundResult의 결과를 Cars 내부에서 한번 더 검증하고 싶었어요.
그래서 getRoundResult()와 makeCarMove() 메서드(랜덤 숫자 로직)를 분리하였답니다.
해당 메서드들을 고민해보니 움직임은 Car 에서 충분히 테스트되었고, RoundResult에 저장된 위치 검증은 요구사항에 없는 과엔지니어링이란걸 깨달았습니다.

> usecase에서도 항상 함께 쓰이고 있는데, 하나의 메서드로 합쳐 제공한다면 시간의존으로 인한 버그 (makeCarMove를 먼저 두번하고 roundResult를 두번 호출한다던지)를 막을 수 있을거 같아요.

이 예시 덕분에 단번에 이해되요! 요청이 동시에 무수히 들어오면 충분히 버그가 생길 거 같아요.
반영하겠습니다!

sihyung92 commented on src/main/java/racingcar/model/RacingGame.java:
```diff
@@ -2,32 +2,22 @@
     }
 
     public TotalResult run() {
         TotalResult totalResult = new TotalResult();
-        for (int i = 0; i < tryCount; i++) {
+        for (int i = 0; i < tryCount.getCount(); i++) {
             cars.makeCarMove(randomNumberGenerator);
             totalResult.addResult(cars.getRoundResult());
         }
         return totalResult;
     }
```
TotalResult와 RoundResult를 생성하는 객체가 통일되었으면 좋겠습니다.

TotalResult는 RoundResult의 컬렉션인데, RoundResult의 생성은 Cars가, TotalResult의 생성은 RacingGame이 맡고 있습니다. 
그래서 TotalResult 객체와 관련해 TotalResult / RacingGame / Cars간의 결합이 있는데요, (TotalResult 생성은 항상 3개 객체가 협력해야함) 
객체의 세계에선 모르는게 약이라서 (최소정보원칙) RacingGame와 Cars가 원만한 합의를 통해 TotalResult + RoundResult 생성 책임을 한 객체로 몰아주면 좋을거 같습니다. 
둘 중 어느 객체가 가져가도 괜찮아보이는데, 개인 취향은 Cars가 몰아가지는 겁니다.

ikjo39 replied:
헉 이게 이론으로만 듣던 결합도 개념이군요. 실제 적용해서 생각해보는건 처음이네요 😮
객체 간의 연관성은 고민하지 못했던 부분인데 좋은 기준점을 배웠어요. 
감사합니다!

sihyung92 replied:
넵 ㅎ 객체지향을 공부하다보면 결합도(coupling)에 대해 자주 듣게 될텐데요, 어떤 경로로든 알게되면(필드로, 파라미터로, 응답값으로..) 결합도가 생겼다고 볼 수 있습니다.

객체지향에선 결합을 가능한 낮추라고 하는데요, 어찌 생각해보면 main문 한개에서 모든 코드를 짜면 결합도가 가장 낮긴 하겠죠.. ㅎ 
제 생각에 핵심은 '낮추라' 보다도 '가능한' 인거 같습니다. 일을 하긴 해야하니까 꼭 필요한 객체들은 알아야 겠죠 ㅎ

좋은 코드에 강박이 있다보면 이런 게 너무 어렵게 느껴지긴 하는데요, 우테코가 강조하는대로 "돌아가는 코드"에 집중하면서 하나씩 고쳐나가는게 방법인거 같습니다

sihyung92 commented on src/main/java/racingcar/controller/RacingController.java:
```diff
@@ -32,14 +34,14 @@ public void run() {
     private Cars createCars() {
         List<String> carNames = inputView.readCarNames();
         List<Car> cars = carNames.stream()
+                .map(CarName::new)
                 .map(Car::new)
                 .toList();
         return new Cars(cars);
```
아래처럼도 가능하겠네요.
```suggestion
        return carNames.stream()
                .map(CarName::new)
                .map(Car::new)
                .collect(Collectors.collectingAndThen(Collectors.toList(), Cars::new));
```

static import를 이용하면 마지막 행의 길이를 좀 더 줄일 수 있습니다
```java
                .collect(collectingAndThen(toList(), Cars::new));
```

ikjo39 replied:
몰랐던 걸 알게되는 쾌감.... 행복합니다 🥰

sihyung92 commented on src/main/java/racingcar/model/Position.java:
```diff
@@ -3,12 +3,17 @@
 import java.util.Objects;
 
 public class Position {
```
> 불필요한 객체 생성에 관련하여 

VO로 만들거나 불변 데이터를 활용하면 필연적으로 메모리를 낭비하는 결과로 이어집니다. 그런데 왜 좋은 코드 규칙에선 VO, 불변 등을 강조할까요?
성능은 당연히 중요하지만 유지보수도 중요한 가치이기 때문입니다. 16kb 칩에 마리오 게임을 구현해야 하던 시절과 달리, 요새는 컴퓨팅 성능과 프로그래밍 언어가 발전하면서 클래스 생성 비용 정도는 시스템 성능에 거의 영향을 주진 않습니다.. (하드웨어 메모리의 증가 및 GC 메커니즘이 향상되었으므로)
그에 비해 버그가 한번 나거나 개발자가 퇴사하여 새로운 개발자가 특정 코드를 이어 작성해야할 때 발생할 비즈니스 비용이나 인건비는 매우 높습니다.

성능을 확보하려는 노력은 항상 중요하지만 다른 factor들도 많으므로, 여러가지를 고려해서 코딩해야합니다. 대부분은 유지보수가 좋은 방향이 훨씬 좋은 성과로 이어집니다. 

성능을 언제쯤 챙기면 될지는 "문제가 될 때까지" 라고 말씀드리겠습니다. 문제가 생기기 전까지 성능은 문제가 아닙니다. 
1ms에 응답하는 API와 50ms에 응답하는 API는 50배의 성능차이가 있지만 비즈니스적으론 아무 상관이 없습니다. 
그러나 1s와 100ms는 10배의 차이이지만 인간이 느낄 수 있으므로 아주 중대한 차이입니다. 앞의 경우엔 개선을 할 필요가 없지만 뒤에 경우엔 성능 개선이 필요하겠죠. (물론 작은 조건들이 쌓여 성능이 나빠지니 아예 성능을 신경쓰지 않는 것은 안 좋지만요)

ikjo39 replied:
아직은 너어어어무 이른 고민이란 말씀이시군요 ㅎ
실생활에서 생각해볼 수 있는 예시는 정말... 감동입니다.
GC 메커니즘이라는 공부 포인트도 얻어가네요 😁

sihyung92 replied:
네네 ㅎ 프로그래밍 언어에서 'pointer'라는 개념을 아실까 모르겠지만 (객체의 메모리 상 주소를 나타내는데 직접 지정해주어야 하는 언어도 있습니다), java에서 pointer라는 개념을 없앤 것도 어찌보면 작은 성능보다는 유지보수에 집중하라는 (넌 프로그램만 짜..) 선배 개발자들의 뜻이기도 할거에요. pointer를 없애면서 메모리에서 객체를 자동으로 지워주는 GC라는 개념도 등장하게 되었는데, java의 발전이 곧 GC 효율성의 발전이기도 합니다. 뭔가 학습 할 때 이렇게 역사(이 개념이 왜 생겼을까?)대로 공부하면 좀더 잘되는거 같기도 하네요.

sihyung92 reviewed (changes requested):
안녕하세요 다온~ 
step1의 리뷰가 잘 반영된 것을 확인했습니당!

Q. 리팩토링 문서 작성 해야하나요?
A. 기능변경 히스토리를 남겨두는 것은 좋은 습관이라고 생각합니다.  선후는 딱히 중요하다고 생각하진 않지만 기능정의서를 작성할 때 처럼 리팩토링할 내용들을 먼저 적어두면 정리도 잘 되고 진행할 때도 편할거라는 생각은 드네요 ㅎ

Q. 현업에서는 어떤 문서를 작성하세요?
A. '작성안함' 일 때도 많았고요(바쁠 땐 ㅎ), 개발 검토를 위한 검토문서, 전체 아키텍처를 한 눈에 볼 수 있는 아키텍처 문서, 플로우 차트가 포함된 개발내역서, 자동 문서화 도구로 설명하기 힘든 내용이 담긴 API문서, 인수인계 문서, 기대효과 문서, 배포 후 성능 측정 문서, 운영업무 대응문서 정도가 생각나네요 

간단한 리뷰 남겼으니 확인해주세용~~

Commit:
Style: 가독성을 위한 변수 분리

Commit:
refactor: Cars 내 항상 함께 쓰이는 메서드 통합

- makeCarMove()
- getRoundResult()

Commit:
refactor: TotalResult 생성 로직 위치 변경

- RacingGame -> Cars
- makeCarsMove() 접근제어자 private 변경
- makeCarMove() 테스트 코드 삭제

Commit:
refactor: Cars 생성 로직 최적화

- Collectors.collectingAndThen() 적용

Commit:
refactor: 불필요한 생성자 삭제

sihyung92 commented on src/main/java/racingcar/model/Cars.java:
```diff
@@ -15,16 +15,12 @@ public Cars(List<Car> cars) {
-        Map<CarName, Position> carStatus = new LinkedHashMap<>();
-        for (Car car : cars) {
-            carStatus.put(car.getCarName(), car.getPosition());
+    public TotalResult getTotalResult(TryCount tryCount, RandomNumberGenerator generator) {
```
이러면 이번엔 TryCount를 알게 되는데, Cars를 원하는 횟수만큼 움직이게 하는건 RacingGame의 일 아닐까요?
제가 원하는 방향은 아래와 같기는 했습니다.
궁극적으로 원했던 건 Cars.makeCarsMove 메서드가 테스트가 가능해지길 원했습니다 ㅎ

```java
    private final List<Car> cars;
    private final TotalResult totalResult = new TotalResult();

...

    public RoundResult makeCarsMove(RandomNumberGenerator generator) {
        Map<CarName, Position> carStatus = new LinkedHashMap<>();
        for (Car car : cars) {
            car.move(generator.generate());
            carStatus.put(car.getCarName(), car.getPosition());
        }
        RoundResult roundResult = new RoundResult(carStatus);
        totalResult.addResult(roundResult);
        return roundResult;
    }

    public TotalResult getTotalResult() {
        return totalResult;
    }
```

sihyung92 replied:
리뷰가 너무 제가 원하는 방향으로 가이딩하는 느낌이어서요, 이걸로 다시 리뷰요청을 보내는 것보단 미션 요구사항 만족도를 보고 머지하는 방향으로 정했습니다 -.-ㅋ
요 방향으로도 한번 확인해보셔요!

ikjo39 replied:
제 코드는 여전히 TryCount 와 결합이 있겠군요 ㅎ
이 방향도 같이 고민해볼게요! 감사합니다

sihyung92 reviewed (approved):
안녕하세요 다온~
첫 미션 재밌으셨으면 좋겠네요 ㅎ 
남은 미션도 꽃밭만 걸으시길 기원하면서 이번 미션은 머지하겠습니다.
짤막한 리뷰 남겼으니 확인해보셔요~
수고하셨습니다!
